pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,418 @@
+		newfssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		if reflect.DeepEqual(newfssConfigMap.Data, oldfssConfigMap.Data) {
+			return
+		}
+		log.Infof("observed fss update: oldfss: %+v newfss: %+v", oldfssConfigMap.Data, newfssConfigMap.Data)
+		var err error
+		var featureStates []featurestatesv1alpha1.FeatureState
+		for feature, state := range newfssConfigMap.Data {
+			var featureState featurestatesv1alpha1.FeatureState
+			featureState.Name = feature
+			featureState.Enabled, err = strconv.ParseBool(state)
+			if err != nil {
+				log.Warnf("failed to parse feature state value: %v for feature: %q", state, feature)
+			}
+			featureStates = append(featureStates, featureState)
+		}
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForAllWorkloadNamespaces(ctx, featureStates)
+	}
+}
+
+// configMapDeleted is called when config-map is deleted
+func configMapDeleted(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	fssConfigMap, ok := obj.(*v1.ConfigMap)
+	if fssConfigMap == nil || !ok {
+		log.Warnf("configMapDeleted: unrecognized object %+v", obj)
+		return
+	}
+	if fssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		fssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		log.Warnf("supervisor feature switch state configmap %q from the namespace: %q is deleted", supervisorFeatureStatConfigMapName, supervisorFeatureStateConfigMapNamespace)
+	}
+}
+
+// namespaceAdded adds is called when new namespace is added on the k8s cluster.
+func namespaceAdded(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	namespace, ok := obj.(*v1.Namespace)
+	if namespace == nil || !ok {
+		log.Warnf("namespaceAdded: unrecognized object %+v", obj)
+		return
+	}
+	if _, ok = namespace.Labels[WorkLoadNamespaceLabelKey]; ok {
+		log.Infof("observed new workload namespace: %v", namespace.Name)
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForWorkloadNamespace(ctx, namespace.Name)
+	}
+}
+
+// namespaceUpdated is called when namespace is updated
+func namespaceUpdated(oldObj, newObj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	oldNamespace, ok := oldObj.(*v1.Namespace)
+	if oldNamespace == nil || !ok {
+		log.Warnf("namespaceUpdated: unrecognized object %+v", oldObj)
+		return
+	}
+	newNamespace, ok := newObj.(*v1.Namespace)
+	if newNamespace == nil || !ok {
+		log.Warnf("namespaceUpdated: unrecognized object %+v", oldObj)
+		return
+	}
+
+	_, labelPresentInOldNamespace := oldNamespace.Labels[WorkLoadNamespaceLabelKey]
+	_, labelPresentInNewNamespace := newNamespace.Labels[WorkLoadNamespaceLabelKey]
+
+	if !labelPresentInOldNamespace && labelPresentInNewNamespace {
```

Comment:
All we care about is whether the label is present in the latest namespace object right?

Reply:
yes so when namespace is updated, if old namespace object does not have workload namespace label, but new namespace object has workload namespace label, means namespace is now being used as workload namespace, so adding it to the map.

for all other updates, `namespaceUpdated` will have no-op.